storage: Add PowerStore driver (#18172)
canonical/lxd: doc/metadata.txt

```

```{config:option} rsync.bwlimit storage-powerstore-pool-conf
:defaultdesc: "`0` (no limit)"
:scope: "global"
:shortdesc: "Upper limit on the socket I/O for `rsync`"
:type: "string"
When `rsync` must be used to transfer storage entities, this option specifies the upper limit
to be placed on the socket I/O.
```

```{config:option} rsync.compression storage-powerstore-pool-conf
:defaultdesc: "`true`"
:scope: "global"
:shortdesc: "Whether to use compression while migrating storage pools"
:type: "bool"

```

```{config:option} volume.size storage-powerstore-pool-conf
:defaultdesc: "`10GiB`"
:scope: "global"
:shortdesc: "Size/quota of the storage volume"
:type: "string"
The size must be in multiples of 1 MiB. The minimum size is 1 MiB and maximum is 256 TiB.
```

<!-- config group storage-powerstore-pool-conf end -->
<!-- config group storage-powerstore-volume-conf start -->
```{config:option} block.filesystem storage-powerstore-volume-conf
:condition: "block-based volume with content type `filesystem`"
:defaultdesc: "same as `volume.block.filesystem`"
:scope: "global"
:shortdesc: "File system of the storage volume"
:type: "string"
Valid options are: `btrfs`, `ext4`, `xfs`

> If not set, `ext4` is assumed.

Reviewer: ```suggestion
Valid options: `btrfs`, `ext4`, `xfs`.
If not set, `ext4` is assumed.
```

```

```{config:option} block.mount_options storage-powerstore-volume-conf
:condition: "block-based volume with content type `filesystem`"
:defaultdesc: "same as `volume.block.mount_options`"
:scope: "global"
:shortdesc: "Mount options for block-backed file system volumes"
:type: "string"

```

```{config:option} security.shared storage-powerstore-volume-conf
:condition: "virtual-machine or custom block volume"
:defaultdesc: "same as `volume.security.shared` or `false`"
:scope: "global"
:shortdesc: "Enable volume sharing"
:type: "bool"

> Enabling this option allows sharing the volume across multiple instances despite the possibility of data loss.

Reviewer: ```suggestion
Enable this option to allow the volume to be shared across multiple instances despite the possibility of data loss.
```

```

```{config:option} security.shifted storage-powerstore-volume-conf
:condition: "custom volume"
:defaultdesc: "same as `volume.security.shifted` or `false`"
:scope: "global"
:shortdesc: "Enable ID shifting overlay"
:type: "bool"

> Enabling this option allows attaching the volume to multiple isolated instances.

Reviewer: ```suggestion
Enable this option to allow the volume to be attached to multiple isolated instances.
```

```

```{config:option} security.unmapped storage-powerstore-volume-conf
:condition: "custom volume"
:defaultdesc: "same as `volume.security.unmapped` or `false`"
:scope: "global"
:shortdesc: "Disable ID mapping for the volume"
:type: "bool"

```

```{config:option} size storage-powerstore-volume-conf
:defaultdesc: "same as `volume.size`"
:scope: "global"
:shortdesc: "Size/quota of the storage volume"
:type: "string"
The size must be in multiples of 1 MiB. The minimum size is 1 MiB and maximum is 256 TiB.
```

```{config:option} snapshots.expiry storage-powerstore-volume-conf
:condition: "custom volume"
:defaultdesc: "same as `volume.snapshots.expiry`"
:scope: "global"
:shortdesc: "When snapshots are to be deleted"
:type: "string"

> Specify an expression like `1M 2H 3d 4w 5m 6y`.

Reviewer: These are different options, right?

```suggestion
Specify an expression such as `1M`, `2H`, `3d`, `4w`, `5m`, or `6y`.
```

Author: They can be combined or used individually. The example shows combined option

Reviewer 2: @MusicDin should this be marked as resolved?

```

```{config:option} snapshots.pattern storage-powerstore-volume-conf
:condition: "custom volume"
:defaultdesc: "same as `volume.snapshots.pattern` or `snap%d`"
:scope: "global"
:shortdesc: "Template for the snapshot name"
:type: "string"
You can specify a naming template that is used for scheduled snapshots and unnamed snapshots.

> The `snapshots.pattern` option takes a Pongo2 template string to format the snapshot name.

Reviewer: ```suggestion
You can specify a naming template for scheduled snapshots and unnamed snapshots.

The `snapshots.pattern` option uses a Pongo2 template string to format the snapshot name.
```

To add a time stamp to the snapshot name, use the Pongo2 context variable `creation_date`.
Make sure to format the date in your template string to avoid forbidden characters in the snapshot name.
For example, set `snapshots.pattern` to `{{ creation_date|date:'2006-01-02_15-04-05' }}` to name the snapshots after their time of creation, down to the precision of a second.

Another way to avoid name collisions is to use the placeholder `%d` in the pattern.
For the first snapshot, the placeholder is replaced with `0`.
For subsequent snapshots, the existing snapshot names are taken into account to find the highest number at the placeholder's position.

> This number is then incremented by one for the new name.

Reviewer: How about simplifying this to:

```suggestion
For each subsequent snapshot, this number is incremented by one.
```

```

```{config:option} snapshots.schedule storage-powerstore-volume-conf
:condition: "custom volume"
:defaultdesc: "same as `snapshots.schedule`"
:scope: "global"
:shortdesc: "Schedule for automatic volume snapshots"
:type: "string"
Specify either a cron expression (`<minute> <hour> <dom> <month> <dow>`), a comma-separated list of schedule aliases (`@hourly`, `@daily`, `@midnight`, `@weekly`, `@monthly`, `@annually`, `@yearly`), or leave empty to disable automatic snapshots (the default).
```

```{config:option} volatile.devlxd.owner storage-powerstore-volume-conf
:defaultdesc: "DevLXD owner identity ID"
:scope: "global"

> :shortdesc: "The ID of the DevLXD identity which owns the volume"

Reviewer: ```suggestion
:shortdesc: "ID of the DevLXD identity that owns the volume"
```

:type: "string"

```

```{config:option} volatile.idmap.last storage-powerstore-volume-conf
:condition: "filesystem"
:shortdesc: "JSON-serialized UID/GID map that has been applied to the volume"
:type: "string"

```

```{config:option} volatile.idmap.next storage-powerstore-volume-conf
:condition: "filesystem"
:shortdesc: "JSON-serialized UID/GID map that has been applied to the volume"
:type: "string"

```

```{config:option} volatile.uuid storage-powerstore-volume-conf
:defaultdesc: "random UUID"
:scope: "global"

> :shortdesc: "The volume's UUID"

Reviewer: A small change, but trying to be consistent with the other options. You typically begin each short description with a noun, rather than a "the."

```suggestion
:shortdesc: "Volume UUID"
```
